libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,622 @@
+{
+    int64_t n_ionized_tot = 0, n_gas_tot = 0;
+    int i, n_ionized = 0;
+    #pragma omp parallel for reduction(+:n_ionized)
+    for (i = 0; i < PartManager->NumPart; i++){
+        if (P[i].Type == 0 && P[i].HeIIIionized == 1){
+            n_ionized ++;
+        }
+    }
+    /* Get total ionization fraction: note this is only the current gas particles.
+     * Particles that become stars are not counted.*/
+    sumup_large_ints(1, &n_ionized, &n_ionized_tot);
+    sumup_large_ints(1, &SlotsManager->info[0].size, &n_gas_tot);
+    return (double) n_ionized_tot / (double) n_gas_tot;
+}
+
+/* Do the ionization for a single particle, marking it and adding the heat.
+ * No locking is performed so ensure the particle is not being edited in parallel.
+ * Returns 1 if ionization was done, 0 otherwise.*/
+static int
+ionize_single_particle(int other)
+{
+    /* Only ionize things once*/
+    if(P[other].HeIIIionized != 0)
+        return 0;
+
+    /* Mark it ionized*/
+    P[other].HeIIIionized = 1;
+
+    /* Heat the particle*/
+
+    /* Number of helium atoms per g in the particle*/
+    double nheperg = (1 - HYDROGEN_MASSFRAC) / (PROTONMASS * HEMASS);
+    /* Total heating per unit mass ergs/g for the particle*/
+    double deltau = qso_inst_heating * nheperg;
+    double uu_in_cgs = All.UnitEnergy_in_cgs / All.UnitMass_in_g;
+
+    /* Conversion factor between internal energy and entropy.*/
+    double entropytou = pow(SPH_EOMDensity(other) * All.cf.a3inv, GAMMA_MINUS1) / GAMMA_MINUS1;
+    /* Convert to entropy in internal units*/
+    SPHP(other).Entropy += deltau / uu_in_cgs / entropytou;
+    return 1;
+}
+
+struct QSOPriv {
+    /* Particle SpinLocks*/
+    struct SpinLocks * spin;
+    FOFGroups * fof;
+    int64_t * N_ionized;
+};
+#define QSO_GET_PRIV(tw) ((struct QSOPriv *) (tw->priv))
+
+/**
+ * Ionize and heat the particles
+ */
+static void
+ionize_ngbiter(TreeWalkQueryBase * I,
+        TreeWalkResultBase * O,
+        TreeWalkNgbIterBase * iter,
+        LocalTreeWalk * lv)
+{
+
+    if(iter->other == -1) {
+        /* Gas only ( 1 == 1 << 0, the bit for type 0)*/
+        iter->mask = 1;
+        /* Bubble size*/
+        double bubble = gaussian_rng(QSOLightupParams.mean_bubble, sqrt(QSOLightupParams.var_bubble), I->ID);
+        iter->Hsml = bubble;
+        /* Don't care about gas HSML */
+        iter->symmetric = NGB_TREEFIND_ASYMMETRIC;
+        return;
+    }
+
+    int other = iter->other;
+
+    /* Only ionize gas*/
+    if(P[other].Type != 0)
+        return;
+
+    struct SpinLocks * spin = QSO_GET_PRIV(lv->tw)->spin;
+
+    lock_spinlock(other, spin);
+
+    int ionized = ionize_single_particle(other);
+
+    unlock_spinlock(other, spin);
+
+    if(!ionized)
+        return;
+
+    int tid = omp_get_thread_num();
+    /* Add to the ionization counter for this thread*/
+    QSO_GET_PRIV(lv->tw)->N_ionized[tid] ++;
+}
+
+static void
+ionize_copy(int place, TreeWalkQueryBase * I, TreeWalk * tw)
```

Comment:
I see. You avoided particle altogether by directly accessnig the fof group.

I never realized tw_copy can be used this way!

Reply:
Exactly! It pays off that you wrote a general interface!